src/System.Security.Cryptography.Encoding/src/System/Security/Cryptography/Base64Transforms.cs

```diff
@@ -96,135 +103,217 @@ public class FromBase64Transform : ICryptoTransform
         public virtual bool CanReuseTransform => true;
 
         public int TransformBlock(byte[] inputBuffer, int inputOffset, int inputCount, byte[] outputBuffer, int outputOffset)
         {
-            ValidateTransformBlock(inputBuffer, inputOffset, inputCount);
-            if (_inputBuffer == null) throw new ObjectDisposedException(null, SR.ObjectDisposed_Generic);
+            // inputCount != InputBlockSize is not allowed
+            ThrowHelper.ValidateTransformBlock(inputBuffer, inputOffset, inputCount, InputBlockSize);
+            if (outputBuffer == null) ThrowHelper.ThrowArgumentNull(ThrowHelper.ExceptionArgument.outputBuffer);
+
+            if (_inputBuffer == null)
+            {
+                ThrowHelper.ThrowObjectDisposed();
+            }
+
+            Debug.Assert(InputBlockSize == 4);
+            Span<byte> tmpBuffer = stackalloc byte[4];
 
-            int effectiveCount;
-            byte[] temp = GetTempBuffer(inputBuffer, inputOffset, inputCount, out effectiveCount);
+            tmpBuffer = GetTempBuffer(inputBuffer.AsSpan(inputOffset, inputCount), tmpBuffer);
+            int bytesToTransform = _inputIndex + tmpBuffer.Length;
 
-            if (effectiveCount + _inputIndex < 4)
+            // To less data to decode: save data to _inputBuffer, so it can be transformed later
+            if (bytesToTransform < InputBlockSize)
             {
-                Buffer.BlockCopy(temp, 0, _inputBuffer, _inputIndex, effectiveCount);
-                _inputIndex += effectiveCount;
+                tmpBuffer.CopyTo(_inputBuffer.AsSpan(_inputIndex));
+
+                _inputIndex = bytesToTransform;
                 return 0;
             }
 
-            byte[] result = ConvertFromBase64(temp, effectiveCount);
+            ConvertFromBase64(tmpBuffer, outputBuffer.AsSpan(outputOffset), out int consumed, out int written);
+            Debug.Assert(consumed == bytesToTransform);
 
-            Buffer.BlockCopy(result, 0, outputBuffer, outputOffset, result.Length);
-
-            return result.Length;
+            return written;
         }
 
         public byte[] TransformFinalBlock(byte[] inputBuffer, int inputOffset, int inputCount)
         {
-            ValidateTransformBlock(inputBuffer, inputOffset, inputCount);
-            if (_inputBuffer == null) throw new ObjectDisposedException(null, SR.ObjectDisposed_Generic);
+            // inputCount > InputBlockSize is allowed
+            ThrowHelper.ValidateTransformBlock(inputBuffer, inputOffset, inputCount);
+
+            if (_inputBuffer == null)
+            {
+                ThrowHelper.ThrowObjectDisposed();
+            }
+
+            if (inputCount == 0)
+            {
+                return Array.Empty<byte>();
+            }
+
+            // The common case is <= 4
+            byte[] tmpBufferArray = null;
+            Span<byte> tmpBuffer = inputCount <= 4
+                ? stackalloc byte[4]
+                : inputCount <= 256
```

Comment:
As the common case is 4, does it make sense to distinguish with 256 before renting from array-pool?

I mean: if it's 4, stackalloc, otherwise rent from pool. So the common case is well covered. Or should I leave this as is?

Reply:
Note: I know that `Span<byte> span = stackalloc byte[4]` could be written `int`-based, yielding slightly better performance. But in context of this method this (little) perf-win doesn't matter, so I went the straight route and hope that the JIT will handle this (anytime) for us.

Reply:
This pair of linked ternaries confused me a little. Perhaps some braces might be clearer?

You may already be aware but it's worth pointing out that in corefx stackalloc does not zero the memory location so you may need to carefully audit the use of suck buffers to make sure the bounds checks are all correct. 

Have you looked at the generated IL/asm for this? It might be allocating both the stackallocs and just swapping in the ref depending on the condition. In general if you want a stackalloc it's easier if it's unconditional so you might want to try out just declaring the 256 buffer at the start and conditionally using it where appropriate.

Reply:
> This pair of linked ternaries 

I think to through out the second ternary, so just stackalloc the common case. That should be clearer thant too.

> It might be allocating both the stackallocs and just swapping in the ref depending on the condition

It jumps to the correct stackallocation, the not-taken branch isn't allocated.

> In general if you want a stackalloc it's easier if it's unconditional

I'll try this. Thanks for this hint.

Reply:
I've update this -> 6087a2c79e0aceeae8dc96036c6809ece093c996
Under the assumption of the common-case with size 4, this results in smaller code with two less branches.